animation_nodes/nodes/spline/splines_from_edges_utils.pyx

```diff
@@ -0,0 +1,114 @@
+    memset(neighboursAmounts, 0, verticesAmount * sizeof(int))
+    for i in range(edgesAmount):
+        neighboursAmounts[edges.data[i].v1] += 1
+        neighboursAmounts[edges.data[i].v2] += 1
+
+    # Compute the start index of each group of neighbours of each vertex.
+    cdef int *neighboursStarts = <int*>PyMem_Malloc(verticesAmount * sizeof(int))
+    cdef int start = 0
+    for i in range(verticesAmount):
+        neighboursStarts[i] = start
+        start += neighboursAmounts[i]
+
+    # Keep track of how many index is in each group of neighbours at each iteration.
+    cdef int *filledSpaces = <int*>PyMem_Malloc(verticesAmount * sizeof(int))
+    memset(filledSpaces, 0, verticesAmount * sizeof(int))
+
+    # Compute the indices of neighbouring vertices of each vertex.
+    cdef int *neighbours = <int*>PyMem_Malloc(edgesAmount * 2 * sizeof(int))
+    for i in range(edgesAmount):
+        neighbours[neighboursStarts[edges.data[i].v1] + filledSpaces[edges.data[i].v1]] = edges.data[i].v2
+        filledSpaces[edges.data[i].v1] += 1
+
+        neighbours[neighboursStarts[edges.data[i].v2] + filledSpaces[edges.data[i].v2]] = edges.data[i].v1
+        filledSpaces[edges.data[i].v2] += 1
+
+    # Find the indices of the vertices that are not connected to two vertices.
+    cdef int *nonBipolarVertices = <int*>PyMem_Malloc(verticesAmount * sizeof(int))
+    cdef int nonBipolarVertsCount = 0
+    for i in range(verticesAmount):
+        if neighboursAmounts[i] != 2:
+            nonBipolarVertices[nonBipolarVertsCount] = i
+            nonBipolarVertsCount += 1
+    nonBipolarVertices = <int*>PyMem_Realloc(nonBipolarVertices, nonBipolarVertsCount * sizeof(int))
+
+    # Generate Splines.
+    cdef list splines = []
+    cdef PolySpline spline
+    cdef FloatList splineRadii
+    cdef Vector3DList splineVertices
+    cdef int nonBipolarVertex, currentVertex, nextVertex
+    for i in range(nonBipolarVertsCount):
+        nonBipolarVertex = nonBipolarVertices[i]
+        for j in range(neighboursAmounts[nonBipolarVertex]):
+            nextVertex = neighbours[neighboursStarts[nonBipolarVertex] + j]
+            if (filledSpaces[nextVertex] == neighboursAmounts[nextVertex] or
+                                           neighboursAmounts[nextVertex] != 2):
+                splineVertices = Vector3DList.__new__(Vector3DList, capacity = verticesAmount)
+                splineRadii = FloatList.__new__(FloatList, capacity = verticesAmount)
+
+                splineVertices.append(vertices[nonBipolarVertex])
+                splineRadii.append(radii.get(nonBipolarVertex))
+                splineVertices.append(vertices[nextVertex])
+                splineRadii.append(radii.get(nextVertex))
+                filledSpaces[nonBipolarVertex] -= 1
+                filledSpaces[nextVertex] -= 1
+
+                currentVertex = nonBipolarVertex
+                while neighboursAmounts[nextVertex] == 2:
+                    i = currentVertex
```

Comment:
the thing is this: it is not used now but you never know if it will be used in the future. By reusing it you sign up for trouble.

Also if anyone wants to convert the code to c, you can't simply reuse the variable, because you would probably end up in an endless loop or so.

Reply:
I see, will make sure not to do that again.

Reply:
How about this form?
```python
while neighboursAmounts[nextVertex] == 2:
    if neighbours[neighboursStarts[nextVertex]] == currentVertex:
        currentVertex = nextVertex
        nextVertex = neighbours[neighboursStarts[nextVertex] + 1]
    else:
        currentVertex = nextVertex
        nextVertex = neighbours[neighboursStarts[nextVertex]]
```
`currentVertex = nextVertex` will be written twice, but we will avoid creating any temporary variables or extra lines of code.

Reply:
I think such inline expression will also work. Which one should we use?
```python
currentVertex, nextVertex = nextVertex, neighbours[neighboursStarts[nextVertex] + neighbours[neighboursStarts[nextVertex]] == currentVertex]
```